### PR TITLE
Automate releasing the extension to VSCode Marketplace from the CI.

### LINKED ISSUE #2: Automate publishing releases
@hokein suggested to trigger this on commits that change the version in `package.json`, which makes sense to me.

### PR DESCRIPTION
**Pull request by hokein**
The idea is to make a new release whenever there is a commit that changes the version in package.json.

Fixes #2 

I verified that the setup works (all steps are executed) in a separate repo. 
Looks like I don't have a permission to add a secret token in `vscode-clangd` repo.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -21,3 +21,28 @@ jobs:
       with:
         source: 'src test'
         extensions: 'ts'
+
+  publish:
+    name: Publish to VSCode marktplace
```
Comment (sam-mccall) [inline .github/workflows/ci.yml:26]: nit: marketplace

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -21,3 +21,28 @@ jobs:
...
+    runs-on: ubuntu-latest
+    if: contains(github.ref, 'master') # publish it only if the push comes from master.
```
Comment (sam-mccall) [inline .github/workflows/ci.yml:28]: I think it'd be nice to isolate this to a different workflow with different triggers, even if it means ignoring the CI.

If we want to run the tests, maybe we should do that explicitly (and redundantly) in the separate workflow, it's not much code.
I'm not sure it makes any sense to block releases on clang-format.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -21,3 +21,28 @@ jobs:
...
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: detect version
+      uses: EndBug/version-check@v1
+      id: check
+      with:
+        diff-search: true # seach the diff content with version
+        file-name: package.json
+        token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (sam-mccall) [inline .github/workflows/ci.yml:39]: you don't need this for a public repository, per the docs

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -21,3 +21,28 @@ jobs:
...
+    needs: build # make sure the build is passed.
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: detect version
+      uses: EndBug/version-check@v1
+      id: check
+      with:
+        diff-search: true # seach the diff content with version
+        file-name: package.json
```
Comment (sam-mccall) [inline .github/workflows/ci.yml:38]: this is the default value

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -21,3 +21,28 @@ jobs:
...
+        token: ${{ secrets.GITHUB_TOKEN }}
+    - name: version bumped
```
Comment (sam-mccall) [inline .github/workflows/ci.yml:40]: I don't think this step is so useful, the following steps run on the same condition, and it's easy to distinguish steps that are run vs skipped in the github UI

File: .github/workflows/ci.yml (hunk 3 above)
Comment (hokein) [reply .github/workflows/ci.yml]: oh, right, I was testing it with a private repo, without this token, it throwed some errors.

File: .github/workflows/auto-release.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      id: check
+      with:
+        diff-search: true # seach the diff content with version
+    - if: steps.check.outputs.changed == 'true'
+      run: npm ci
+    - if: steps.check.outputs.changed == 'true'
+      run: npm test
+    - name: publish
+      if: steps.check.outputs.changed == 'true'
+      run: npm run publish -p ${{ secrets.VSCODE_MARKETPLACE_TOKEN }}
```
Comment (sam-mccall) [inline .github/workflows/auto-release.yml:25]: nit: quote the token

**Comment by hokein** [comment]
I think it should work, but I didn't use it to make a real release (shall we bump a version for testing?)

Just a note: the secret is added, but it will be expired in on 12/31/2020. Regenerate it at https://llvm-vs-code-extensions.visualstudio.com/_usersSettings/tokens when expired.

**Comment by sam-mccall** [comment]
A real release would be good actually - e.g. for 0e41bf23184f213b317ccd03fe28af9efd2c95c1

So merge this and try to make a release?

> Just a note: the secret is added, but it will be expired in on 12/31/2020. Regenerate it at https://llvm-vs-code-extensions.visualstudio.com/_usersSettings/tokens when expired.

Can you add this as a comment to the yaml workflow?

(BTW this is going to bite us the first time we try to push a release after it expires: the workflow will fail and we don't have a way to force it to rerun other than bumping the version again. Which is probably a fine workaround)

**Comment by hokein** [comment]
> A real release would be good actually - e.g. for [0e41bf2](https://github.com/clangd/vscode-clangd/commit/0e41bf23184f213b317ccd03fe28af9efd2c95c1)
> 
> So merge this and try to make a release?

sounds good.

> Can you add this as a comment to the yaml workflow?

Done.

> (BTW this is going to bite us the first time we try to push a release after it expires: the workflow will fail and we don't have a way to force it to rerun other than bumping the version again. Which is probably a fine workaround)

yeah, we could do `git push -f` :), or setup an alarm when the token is going to expired.

**Pull request merged**